agent/consul/rpc.go

```diff
@@ -434,5 +434,31 @@ func (s *Server) setQueryMeta(m *structs.QueryMeta) {
 func (s *Server) consistentRead() error {
 	defer metrics.MeasureSince([]string{"consul", "rpc", "consistentRead"}, time.Now())
 	future := s.raft.VerifyLeader()
-	return future.Error()
+	err := future.Error()
+	if err != nil {
+		return err //fail fast if leader verification fails
+	}
+	// poll consistent read readiness, wait for up to RPCHoldTimeout milliseconds
+	if s.isReadyForConsistentReads() {
+		return nil
+	}
+	jitter := lib.RandomStagger(s.config.RPCHoldTimeout / jitterFraction)
+	deadline := time.Now().Add(jitter)
+
+	for time.Now().After(deadline) {
```

Comment:
You could cook up a unit test to exercise this where you wait for a leader and make sure that it reports ready for consistent reads and then call the function to reset this from the unit test. You could call this and make sure you get the error you expect saying it's not ready for consistent reads. You could then arrange for a goroutine to sleep for a bit and then set it back to 1, and have the test call this and make sure it eventually unblocks and returns no error.